Client/Frontend/Browser/TopTabsViewController.swift

```diff
@@ -0,0 +1,714 @@
+    }
+    
+    func privateTabClicked() {
+        delegate.topTabsPressPrivateTab()
+        self.collectionView.reloadData()
+        scrollToCurrentTab(false)
+    }
+    
+    func closeTab() {
+        delegate.topTabsPressTabs()
+    }
+    
+    func reloadFavicons() {
+        self.collectionView.reloadData()
+    }
+    
+    func scrollToCurrentTab(animated: Bool = true) {
+        guard let currentTab = tabManager.selectedTab, let index = tabsToDisplay.indexOf(currentTab) else {
+            return
+        }
+        guard !collectionView.frame.isEmpty else {
+            return
+        }
+        if let frame = collectionView.layoutAttributesForItemAtIndexPath(NSIndexPath(forRow: index, inSection: 0))?.frame {
+            // Padding is added to ensure the tab is completely visible (none of the tab is under the fader)
+            let padFrame = frame.insetBy(dx: -(TopTabsUX.TopTabsBackgroundShadowWidth+TopTabsUX.FaderPading), dy: 0)
+            collectionView.scrollRectToVisible(padFrame, animated: true)
+        }
+    }
+    
+}
+
+extension TopTabsViewController: Themeable {
+    func applyTheme(themeName: String) {
+        tabsButton.applyTheme(themeName)
+        if themeName == Theme.PrivateMode {
+            isPrivate = true
+        }
+        else if themeName == Theme.NormalMode {
+            isPrivate = false
+        }
+    }
+}
+
+extension TopTabsViewController: TopTabCellDelegate {
+    func tabCellDidClose(cell: TopTabCell) {
+        let indexPath = collectionView.indexPathForCell(cell)!
+        let tab = tabsToDisplay[indexPath.item]
+        if tab == tabManager.selectedTab {
+            delegate.topTabsDidChangeTab()
+        }
+        if tabsToDisplay.count == 1 {
+            tabManager.removeTab(tab)
+            tabManager.selectTab(tabsToDisplay.first)
+            collectionView.reloadData()
+        }
+        else {
+            tabManager.removeTab(tab)
+            self.collectionView.performBatchUpdates({
+                self.collectionView.deleteItemsAtIndexPaths([indexPath])
+                }, completion: { finished in
+                    if #available(iOS 9, *) {
+                        self.collectionView.reloadData()
+                    }
+            })
+        }
+    }
+}
+
+extension TopTabsViewController: UICollectionViewDataSource {
+    @objc func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
+        let index = indexPath.item
+        let tabCell = collectionView.dequeueReusableCellWithReuseIdentifier(TopTabCell.Identifier, forIndexPath: indexPath) as! TopTabCell
+        tabCell.delegate = self
+        
+        let tab = tabsToDisplay[index]
+        tabCell.style = tab.isPrivate ? .Dark : .Light
+        tabCell.titleText.text = tab.displayTitle
+        
+        if tab.displayTitle.isEmpty {
+            if (tab.url?.baseDomain()?.contains("localhost") ?? false || tab.url == nil) {
```

Comment:
Nit: Don't think the nil check is necessary at the end since the optional will fail in the first expression and result in false.